Plugins/ServerPlugin.cpp

```diff
@@ -199,43 +210,33 @@ void CompilerClient::SyntaxCheck() {
 
   auto worker = dynamic_cast<AsyncResponseReadWorker<SyntaxError>*>(callData);
   worker->stream = stub->PrepareAsyncSyntaxCheck(&worker->context, syntaxCheckRequest, &cq);
+  callData->start();
 }
 
 template <typename T>
 T* CompilerClient::ScheduleTask() {
-  std::unique_ptr<T> callData(new T());
-  tasks.push(std::move(callData));
-  return (T*)tasks.back().get();
+  auto callData = new T();
+  connect(callData, &CallData::LogOutput, this, &CompilerClient::LogOutput);
+  connect(callData, &CallData::CompileStatusChanged, this, &CompilerClient::CompileStatusChanged);
+  return callData;
 }
 
 void CompilerClient::UpdateLoop() {
-  static bool started = false;
   void* got_tag = nullptr;
   bool ok = false;
 
-  if (this->tasks.empty()) return;

```

Comment:
Yeah I was stuffing the tasks in queue before this problem was solved. That way I could just refer to the task/calldata at the front of the queue. In other words, it was synchronous.

If you're worried about resource leaks though, I did make the workers QObject based (so they can use thread-safe signals to update UI). So instead of deleting the calldata when it's finished like Google does, I could actually parent the workers to server plugin itself (also a QObject). Otherwise I could use unique_ptr as the tag instead of the raw pointer.